dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java

```diff
@@ -79,53 +79,66 @@ public CheckpointImpl(ComponentContext cc) {
+        SnapshotResult result = doSnapshot(phase,
+                                           locAdmin.resolveResource(WsLocationConstants.SYMBOL_SERVER_WORKAREA_DIR + "checkpoint/image/").asFile());
+        return result;
     }
 
-    private void doSnapshot(Phase phase, File directory) throws SnapshotFailed {
+    private SnapshotResult doSnapshot(Phase phase, File directory) {
         directory.mkdirs();
         Object[] factories = cc.locateServices("hookFactories");
         List<SnapshotHook> snapshotHooks = getHooks(factories, phase);
-        prepare(snapshotHooks);
-        try {
-            ExecuteCRIU currentCriu = criu;
-            if (currentCriu == null) {
-                throw new SnapshotFailed(Type.SNAPSHOT_FAILED, "The criu command is not available.", null, 0);
-            } else {
-                if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-                    Tr.debug(tc, "criu attempt dump to '" + directory + "' and exit process.");
-                }
+        SnapshotResult prepareResult = prepare(snapshotHooks);
 
-                int dumpCode = currentCriu.dump(directory);
-                if (dumpCode < 0) {
-                    throw new SnapshotFailed(Type.SNAPSHOT_FAILED, "The criu dump command failed with error: " + dumpCode, null, dumpCode);
-                }
+        if (prepareResult.getType() != SnapshotResultType.SUCCESS) {
+            return prepareResult;
+        }
 
-                if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-                    Tr.debug(tc, "criu dump to " + directory + " in recovery.");
-                }
+        ExecuteCRIU currentCriu = criu;
+        SnapshotResult dumpResult = null;
+
+        if (currentCriu == null) {
+            return new SnapshotResult(SnapshotResultType.SNAPSHOT_FAILED, "The criu command is not available.", null);
+        } else {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "criu attempt dump to '" + directory + "' and exit process.");
+            }
+
+            dumpResult = currentCriu.dump(directory);
+
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "criu dump to " + directory + " in recovery.");
             }
-        } catch (Exception e) {
-            abortPrepare(snapshotHooks, e);
-            if (e instanceof SnapshotFailed) {
-                throw (SnapshotFailed) e;
+
+            if (dumpResult.getType() == SnapshotResultType.SUCCESS) {
+                SnapshotResult restoreResult = restore(phase, snapshotHooks);
+                if (restoreResult.getType() != SnapshotResultType.SUCCESS) {
+                    return restoreResult;
+                }
+            } else if (dumpResult.getType() == SnapshotResultType.JVM_RESTORE_FAILURE) {
+                return new SnapshotResult(SnapshotResultType.JVM_RESTORE_FAILURE, "Unexpected JVM failure.", null);
```

Comment:
Why create a new SnapshotResult instead of returning the `dumpResult` directly here?